awx/main/tests/functional/test_inventory_source_injectors.py

```diff
@@ -192,7 +193,7 @@ def create_reference_data(source_dir, env, content):
 
 
 @pytest.mark.django_db
-@pytest.mark.parametrize('this_kind', CLOUD_PROVIDERS)
+@pytest.mark.parametrize('this_kind', discover_available_cloud_provider_plugin_names())
```

Comment:
Looking into the test failures, it seems to me that my suspicion was likely right and this might need to be moved away from import-time invocation. And I'd try converting this to use `pytest-subtests`.

Reply:
@djyasin did this end up working as is? Have you checked that individual tests show up in the `pytest -vv -k test_inventory_update_injected_content` output?

Reply:
That test did end up passing successfully after Chris helped me tweak a few things in our functionality. This is the output of `pytest -vv -k test_inventory_update_injected_content `

`bash-5.1$ pytest -vv -k test_inventory_update_injected_content
/var/lib/awx/venv/awx/lib64/python3.11/site-packages/pytest_asyncio/plugin.py:208: PytestDeprecationWarning: The configuration option "asyncio_default_fixture_loop_scope" is unset.
The event loop scope for asynchronous fixtures will default to the fixture caching scope. Future versions of pytest-asyncio will default the loop scope for asynchronous fixtures to function scope. Set the default fixture loop scope explicitly in order to avoid unexpected behavior in the future. Valid fixture loop scopes are: "function", "class", "module", "package", "session"

  warnings.warn(PytestDeprecationWarning(_DEFAULT_FIXTURE_LOOP_SCOPE_UNSET))
============================================================= test session starts ==============================================================
platform linux -- Python 3.11.10, pytest-8.3.3, pluggy-1.5.0 -- /var/lib/awx/venv/awx/bin/python3.11
cachedir: .pytest_cache
django: version: 4.2.10, settings: awx.main.tests.settings_for_test (from ini)
rootdir: /awx_devel
configfile: pytest.ini
plugins: anyio-4.6.2.post1, django-test-migrations-1.4.0, asyncio-0.24.0, cov-5.0.0, django-4.9.0, mock-3.14.0, timeout-2.3.1, xdist-3.6.1
asyncio: mode=Mode.STRICT, default_loop_scope=None
collected 3308 items / 124 errors                                                                                                              
WARNING setuptools_scm.pyproject_reading toml section missing 'pyproject.toml does not contain a tool.setuptools_scm section'
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/setuptools/_distutils/fancy_getopt.py", line 246, in getopt
INTERNALERROR>     opts, args = getopt.getopt(args, short_opts, self.long_opts)
INTERNALERROR>                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/usr/lib64/python3.11/getopt.py", line 95, in getopt
INTERNALERROR>     opts, args = do_shorts(opts, args[0][1:], shortopts, args[1:])
INTERNALERROR>                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/usr/lib64/python3.11/getopt.py", line 195, in do_shorts
INTERNALERROR>     if short_has_arg(opt, shortopts):
INTERNALERROR>        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/usr/lib64/python3.11/getopt.py", line 211, in short_has_arg
INTERNALERROR>     raise GetoptError(_('option -%s not recognized') % opt, opt)
INTERNALERROR> getopt.GetoptError: option -k not recognized
INTERNALERROR> 
INTERNALERROR> During handling of the above exception, another exception occurred:
INTERNALERROR> 
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/setuptools/_distutils/core.py", line 172, in setup
INTERNALERROR>     ok = dist.parse_command_line()
INTERNALERROR>          ^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/setuptools/_distutils/dist.py", line 467, in parse_command_line
INTERNALERROR>     args = parser.getopt(args=self.script_args, object=self)
INTERNALERROR>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/setuptools/_distutils/fancy_getopt.py", line 248, in getopt
INTERNALERROR>     raise DistutilsArgError(msg)
INTERNALERROR> distutils.errors.DistutilsArgError: option -k not recognized
INTERNALERROR> 
INTERNALERROR> During handling of the above exception, another exception occurred:
INTERNALERROR> 
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/_pytest/main.py", line 283, in wrap_session
INTERNALERROR>     session.exitstatus = doit(config, session) or 0
INTERNALERROR>                          ^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/_pytest/main.py", line 336, in _main
INTERNALERROR>     config.hook.pytest_collection(session=session)
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/pluggy/_hooks.py", line 513, in __call__
INTERNALERROR>     return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
INTERNALERROR>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/pluggy/_manager.py", line 120, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
INTERNALERROR>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/pluggy/_callers.py", line 139, in _multicall
INTERNALERROR>     raise exception.with_traceback(exception.__traceback__)
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/pluggy/_callers.py", line 122, in _multicall
INTERNALERROR>     teardown.throw(exception)  # type: ignore[union-attr]
INTERNALERROR>     ^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/_pytest/logging.py", line 790, in pytest_collection
INTERNALERROR>     return (yield)
INTERNALERROR>             ^^^^^
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/pluggy/_callers.py", line 122, in _multicall
INTERNALERROR>     teardown.throw(exception)  # type: ignore[union-attr]
INTERNALERROR>     ^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/_pytest/warnings.py", line 121, in pytest_collection
INTERNALERROR>     return (yield)
INTERNALERROR>             ^^^^^
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/pluggy/_callers.py", line 122, in _multicall
INTERNALERROR>     teardown.throw(exception)  # type: ignore[union-attr]
INTERNALERROR>     ^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/_pytest/config/__init__.py", line 1417, in pytest_collection
INTERNALERROR>     return (yield)
INTERNALERROR>             ^^^^^
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/pluggy/_callers.py", line 103, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>           ^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/_pytest/main.py", line 347, in pytest_collection
INTERNALERROR>     session.perform_collect()
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/_pytest/main.py", line 809, in perform_collect
INTERNALERROR>     self.items.extend(self.genitems(node))
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/_pytest/main.py", line 975, in genitems
INTERNALERROR>     yield from self.genitems(subnode)
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/_pytest/main.py", line 975, in genitems
INTERNALERROR>     yield from self.genitems(subnode)
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/_pytest/main.py", line 970, in genitems
INTERNALERROR>     rep, duplicate = self._collect_one_node(node, handle_dupes)
INTERNALERROR>                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/_pytest/main.py", line 835, in _collect_one_node
INTERNALERROR>     rep = collect_one_node(node)
INTERNALERROR>           ^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/_pytest/runner.py", line 567, in collect_one_node
INTERNALERROR>     rep: CollectReport = ihook.pytest_make_collect_report(collector=collector)
INTERNALERROR>                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/pluggy/_hooks.py", line 513, in __call__
INTERNALERROR>     return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
INTERNALERROR>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/pluggy/_manager.py", line 120, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
INTERNALERROR>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/pluggy/_callers.py", line 139, in _multicall
INTERNALERROR>     raise exception.with_traceback(exception.__traceback__)
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/pluggy/_callers.py", line 122, in _multicall
INTERNALERROR>     teardown.throw(exception)  # type: ignore[union-attr]
INTERNALERROR>     ^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/_pytest/capture.py", line 860, in pytest_make_collect_report
INTERNALERROR>     rep = yield
INTERNALERROR>           ^^^^^
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/pluggy/_callers.py", line 103, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>           ^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/_pytest/runner.py", line 391, in pytest_make_collect_report
INTERNALERROR>     call = CallInfo.from_call(
INTERNALERROR>            ^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/_pytest/runner.py", line 341, in from_call
INTERNALERROR>     result: TResult | None = func()
INTERNALERROR>                              ^^^^^^
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/_pytest/runner.py", line 389, in collect
INTERNALERROR>     return list(collector.collect())
INTERNALERROR>                 ^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/pytest_asyncio/plugin.py", line 667, in _patched_collect
INTERNALERROR>     module = collector.obj
INTERNALERROR>              ^^^^^^^^^^^^^
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/_pytest/python.py", line 284, in obj
INTERNALERROR>     self._obj = obj = self._getobj()
INTERNALERROR>                       ^^^^^^^^^^^^^^
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/_pytest/python.py", line 546, in _getobj
INTERNALERROR>     return importtestmodule(self.path, self.config)
INTERNALERROR>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/_pytest/python.py", line 493, in importtestmodule
INTERNALERROR>     mod = import_path(
INTERNALERROR>           ^^^^^^^^^^^^
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/_pytest/pathlib.py", line 582, in import_path
INTERNALERROR>     importlib.import_module(module_name)
INTERNALERROR>   File "/usr/lib64/python3.11/importlib/__init__.py", line 126, in import_module
INTERNALERROR>     return _bootstrap._gcd_import(name[level:], package, level)
INTERNALERROR>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
INTERNALERROR>   File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
INTERNALERROR>   File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
INTERNALERROR>   File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/_pytest/assertion/rewrite.py", line 184, in exec_module
INTERNALERROR>     exec(co, module.__dict__)
INTERNALERROR>   File "/awx_devel/awxkit/setup.py", line 78, in <module>
INTERNALERROR>     setup(
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/setuptools/__init__.py", line 103, in setup
INTERNALERROR>     return distutils.core.setup(**attrs)
INTERNALERROR>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INTERNALERROR>   File "/var/lib/awx/venv/awx/lib64/python3.11/site-packages/setuptools/_distutils/core.py", line 174, in setup
INTERNALERROR>     raise SystemExit(gen_usage(dist.script_name) + "\nerror: %s" % msg)
INTERNALERROR> SystemExit: usage: pytest [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
INTERNALERROR>    or: pytest --help [cmd1 cmd2 ...]
INTERNALERROR>    or: pytest --help-commands
INTERNALERROR>    or: pytest cmd --help
INTERNALERROR> 
INTERNALERROR> error: option -k not recognized
mainloop: caught unexpected SystemExit!
`

Reply:
I tweaked what I was running and got successful output. I ran `var/lib/awx/venv/awx/bin/pytest --showlocals -v awx/main/tests/functional/test_inventory_source_injectors.py -k test_inventory_update_injected_content `
And got:
> bash-5.1$ /var/lib/awx/venv/awx/bin/pytest --showlocals -v awx/main/tests/functional/test_inventory_source_injectors.py -k test_inventory_update_injected_content 
/var/lib/awx/venv/awx/lib64/python3.11/site-packages/pytest_asyncio/plugin.py:208: PytestDeprecationWarning: The configuration option "asyncio_default_fixture_loop_scope" is unset.
The event loop scope for asynchronous fixtures will default to the fixture caching scope. Future versions of pytest-asyncio will default the loop scope for asynchronous fixtures to function scope. Set the default fixture loop scope explicitly in order to avoid unexpected behavior in the future. Valid fixture loop scopes are: "function", "class", "module", "package", "session"

  warnings.warn(PytestDeprecationWarning(_DEFAULT_FIXTURE_LOOP_SCOPE_UNSET))
============================================================= test session starts ==============================================================
platform linux -- Python 3.11.10, pytest-8.3.3, pluggy-1.5.0 -- /var/lib/awx/venv/awx/bin/python3.11
cachedir: .pytest_cache
django: version: 4.2.10, settings: awx.main.tests.settings_for_test (from ini)
rootdir: /awx_devel
configfile: pytest.ini
plugins: anyio-4.6.2.post1, django-test-migrations-1.4.0, asyncio-0.24.0, cov-5.0.0, django-4.9.0, mock-3.14.0, timeout-2.3.1, xdist-3.6.1
asyncio: mode=Mode.STRICT, default_loop_scope=None
collected 17 items                                                                                                                             

awx/main/tests/functional/test_inventory_source_injectors.py::test_inventory_update_injected_content[azure_rm] PASSED                    [  5%]
awx/main/tests/functional/test_inventory_source_injectors.py::test_inventory_update_injected_content[ec2] PASSED                         [ 11%]
awx/main/tests/functional/test_inventory_source_injectors.py::test_inventory_update_injected_content[gce] PASSED                         [ 17%]
awx/main/tests/functional/test_inventory_source_injectors.py::test_inventory_update_injected_content[vmware] PASSED                      [ 23%]
awx/main/tests/functional/test_inventory_source_injectors.py::test_inventory_update_injected_content[openstack] PASSED                   [ 29%]
awx/main/tests/functional/test_inventory_source_injectors.py::test_inventory_update_injected_content[rhv] PASSED                         [ 35%]
awx/main/tests/functional/test_inventory_source_injectors.py::test_inventory_update_injected_content[rhv_supported] PASSED               [ 41%]
awx/main/tests/functional/test_inventory_source_injectors.py::test_inventory_update_injected_content[satellite6] PASSED                  [ 47%]
awx/main/tests/functional/test_inventory_source_injectors.py::test_inventory_update_injected_content[satellite6_supported] PASSED        [ 52%]
awx/main/tests/functional/test_inventory_source_injectors.py::test_inventory_update_injected_content[terraform] PASSED                   [ 58%]
awx/main/tests/functional/test_inventory_source_injectors.py::test_inventory_update_injected_content[controller] PASSED                  [ 64%]
awx/main/tests/functional/test_inventory_source_injectors.py::test_inventory_update_injected_content[controller_supported] PASSED        [ 70%]
awx/main/tests/functional/test_inventory_source_injectors.py::test_inventory_update_injected_content[insights] PASSED                    [ 76%]
awx/main/tests/functional/test_inventory_source_injectors.py::test_inventory_update_injected_content[insights_supported] PASSED          [ 82%]
awx/main/tests/functional/test_inventory_source_injectors.py::test_inventory_update_injected_content[openshift_virtualization] PASSED    [ 88%]
awx/main/tests/functional/test_inventory_source_injectors.py::test_inventory_update_injected_content[openshift_virtualization_supported] PASSED [ 94%]
awx/main/tests/functional/test_inventory_source_injectors.py::test_inventory_update_injected_content[constructed] SKIPPED (Use of in...) [100%]